### PR TITLE
Add server sent events

### PR DESCRIPTION
**Pull request by NicolasPetton**
This is a follow-up to the original PR by @Arlen22 here: https://github.com/Jermolene/TiddlyWiki5/pull/4609.

---
#### Original PR description

This is related to Issue #4608 and also fixes Issue #4605.

To test this, run node tiddlywiki.js editions/server --listen then open two pages side-by-side and watch changes propagate in almost real time.

### CODE REVIEW

**Comment by pmario** [comment]
@Jermolene ... bump

**Comment by NicolasPetton** [comment]
@Jermolene are there any changes you'd like me to do?

**Comment by NicolasPetton** [comment]
@Jermolene I just rebased the branch. 

I can also rerwite the history if that's too many commits (I kept the commits from @Arlen22 but I can squash some of them to make the history cleaner).

File: core/modules/server/server-sent-events.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  return {
+    bodyFormat: "stream",
+    method: "GET",
+    path: new RegExp("^/events/" + this.prefix + "$"),
+    handler: this.handleEventRequest.bind(this)
+  };
+};
+
+ServerSentEvents.prototype.handleEventRequest = function(request,response,state) {
+  if(request.headers.accept && request.headers.accept.startsWith("text/event-stream")) {
```
Comment (Jermolene) [inline core/modules/server/server-sent-events.js:36]: `startsWith` is an ECMAScript 2015 thing, we try to stick to 5.1.

File: core/modules/server/server-sent-events.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+parameters:
+		prefix - usually the plugin path, such as `plugins/tiddlywiki/tiddlyweb`. The
+			route will match `/events/${prefix}` exactly.
+
+		handler - a function that will be called each time a request comes in with the
+			request and state from the route and an emit function to call.
+*/
+
+var ServerSentEvents = function ServerSentEvents(prefix, handler) {
+  this.handler = handler;
```
Comment (Jermolene) [inline core/modules/server/server-sent-events.js:22]: We use tabs with a width of 4 for indenting, it looks like these files are indented with spaces.

File: plugins/tiddlywiki/tiddlyweb/sse-server.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+*/
+function setupWiki(wiki) {
+	var index = wikis.length;
+	// Add a new array for this wiki (object references work as keys)
+	wikis.push(wiki);
+	connections.push([]);
+	// Listen to change events for this wiki
+	wiki.addEventListener("change",function(changes) {
+		getWikiConnections(wiki).forEach(function(item) {
+			item.emit("change",JSON.stringify(changes));
```
Comment (Jermolene) [inline plugins/tiddlywiki/tiddlyweb/sse-server.js:29]: I'd be inclined to take the JSON.stringify out of the loop.

**Review by Jermolene — COMMENTED** [review]
Hi @NicolasPetton apologies for the delay. Although I've been merging new stuff as it pops up, I haven't yet had a chance to methodically go back through the tickets we identified in #5231.

There's a couple of minor coding style comments. I'm also wondering if we need a configuration setting to turn off the SSE server; it could conceivably be a vector for DDOS attacks. We also need documentation updates.

Finally, I'm wondering if it's worth merging this without updating the syncer/syncadaptor to take advantage of it.

**Comment by NicolasPetton** [comment]
> There's a couple of minor coding style comments. 

I'll fix them.

> I'm also wondering if we need a configuration setting to turn off the SSE server; it could conceivably be a vector for DDOS attacks. 

Would you accept that improvement in a separate PR?

> We also need documentation updates.

Would you like documentation updates as part of this PR? If so, could you point out which tiddlers need updating?

> Finally, I'm wondering if it's worth merging this without updating the syncer/syncadaptor to take advantage of it.

Sorry, I'm not following. The PR adds a route `/events/plugins/tiddlywiki/tiddlyweb` for tiddlyweb SSE in `sse-server.js`. Have I missed something else?

**Comment by pmario** [comment]
@Jermolene from the OP

>To test this, run node tiddlywiki.js editions/server --listen then open two pages side-by-side and watch changes propagate in almost real time.

I think it does take advantage of the new funtionality

**Comment by pmario** [comment]
I think the configuration may be a new parameter to the `--listen` command see: https://tiddlywiki.com/#ListenCommand

**Comment by Jermolene** [comment]
> I'll fix them.

Thanks.

> Would you accept that improvement in a separate PR?
> 
> Would you like documentation updates as part of this PR? If so, could you point out which tiddlers need updating?

It's generally best to include the docs in the same PR but we do fairly often merge them separately.
 
> > Finally, I'm wondering if it's worth merging this without updating the syncer/syncadaptor to take advantage of it.
> 
> Sorry, I'm not following. The PR adds a route `/events/plugins/tiddlywiki/tiddlyweb` for tiddlyweb SSE in `sse-server.js`. Have I missed something else?

The integration with the syncer is primitive; the client calls `$tw.syncer.syncFromServer()` when it receives changes from the server, but we don't do anything to disable the usual polling based syncing.

**Comment by Jermolene** [comment]
> I think the configuration may be a new parameter to the `--listen` command see: https://tiddlywiki.com/#ListenCommand
> 
> For compatibility and security reasons, it may be switched OFF by default. So if the new parameter is present it will be switched on

That makes sense. The client will need to be able to run correctly with either setting.

File: core/modules/server/server-sent-events.js (hunk 1 above)
Comment (NicolasPetton) [reply core/modules/server/server-sent-events.js]: Done.

File: core/modules/server/server-sent-events.js (hunk 2 above)
Comment (NicolasPetton) [reply core/modules/server/server-sent-events.js]: Fixed.

File: plugins/tiddlywiki/tiddlyweb/sse-server.js (hunk 3 above)
Comment (NicolasPetton) [reply plugins/tiddlywiki/tiddlyweb/sse-server.js]: Done as well.

**Comment by NicolasPetton** [comment]
@Jermolene I've fixed all the styling issues, plus:

- Added a new WebServer parameter `sse-enabled`, which is disabled by default.
- When enabled, polling is disabled in `syncer.js`. The `/status` JSON response now contains a new `sse_enabled` field.
- I documented the new `sse-enabled` param in a new tiddler and added a list item in `core/language/en-GB/Help/listen.tid`

**Review by pmario — COMMENTED** [review]
The code looks good to me

**Comment by Jermolene** [comment]
Bravo, and thank you @NicolasPetton looks great.

**Comment by linonetwo** [comment]
Wow, amazing feature, will this make https://github.com/Jermolene/TiddlyWiki5/issues/3060 possible?

**Comment by Jermolene** [comment]
> Wow, amazing feature, will this make https://github.com/Jermolene/TiddlyWiki5/issues/3060 possible?

Sadly no: #3060 is about syncing between the serverside wiki and the file system, while SSE just improve things between the clientside wiki and the serverside wiki.

**Comment by Jermolene** [comment]
As discussed in https://github.com/Jermolene/TiddlyWiki5/pull/5853#issuecomment-874634324, I'm reverting this PR in favour of a slightly different approach: we will add the hooks required to enable SSE implementations to be undertaken in plugins.

**Comment by Jermolene** [comment]
Apologies, to clarify that the reverting is being done in #5853

**Pull request merged**